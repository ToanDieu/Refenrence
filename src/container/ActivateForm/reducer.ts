import { ActivateFormState, ActivateFormActions } from "./types";
import {
  CLEANUP_NOTIFICATIONS,
  SET_DIALOG_VISIBLE,
  ACTIVATE_WORKFLOW_PENDING,
  ACTIVATE_WORKFLOW_SUCCESS,
  ACTIVATE_WORKFLOW_ERROR
} from "./constants";

export const initialState: ActivateFormState = {
  visible: false,
  submitting: false,
  error: undefined,
  activated: false
};

export default function(
  state = initialState,
  action: ActivateFormActions
): ActivateFormState {
  switch (action.type) {
    case CLEANUP_NOTIFICATIONS:
      return { ...state, error: undefined, activated: false };
    case SET_DIALOG_VISIBLE:
      return { ...state, visible: action.visible };
    case ACTIVATE_WORKFLOW_PENDING:
      return { ...state, submitting: true, error: undefined };
    case ACTIVATE_WORKFLOW_SUCCESS:
      return { ...state, submitting: false, activated: true, visible: false };
    case ACTIVATE_WORKFLOW_ERROR:
      return { ...state, submitting: false, error: action.error };
    default:
      return state;
  }
}
