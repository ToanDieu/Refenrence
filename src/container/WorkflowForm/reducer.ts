import { CREATE_WORKFLOW_SUCCESS } from "@/resources/workflows/constants";
import { WorkflowsActions } from "@/resources/workflows/types";
import { WorkflowFormState, WorflowFormActions } from "./types";
import {
  SET_DIALOG_VISIBLE,
  CHANGE_FORM_DATA,
  CREATE_WORKFLOW_PENDING,
  CREATE_WORKFLOW_ERROR
} from "./constants";

export const initialForm = {
  name: "",
  description: ""
};

export const initialState: WorkflowFormState = {
  visible: false,
  submitting: false,
  error: undefined,
  form: initialForm
};

export default function(
  state = initialState,
  action: WorflowFormActions | WorkflowsActions
): WorkflowFormState {
  switch (action.type) {
    case SET_DIALOG_VISIBLE:
      return { ...state, visible: action.visible };
    case CHANGE_FORM_DATA:
      return {
        ...state,
        form: {
          ...state.form,
          [action.key]: action.value
        }
      };
    case CREATE_WORKFLOW_PENDING:
      return { ...state, submitting: true, error: undefined };
    case CREATE_WORKFLOW_SUCCESS:
      return { ...state, submitting: false, form: initialForm, visible: false };
    case CREATE_WORKFLOW_ERROR:
      return { ...state, submitting: false, error: action.error };
    default:
      return state;
  }
}
