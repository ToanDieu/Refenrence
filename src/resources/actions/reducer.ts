import { combineReducers } from "redux";
import { set, union } from "lodash/fp";
import { dissocPath, pathOr, assocPath } from "ramda";

import {
  LOAD_WORKFLOW_ACTIONS_SUCCESS,
  CREATE_WORKFLOW_TEMP_ACTION,
  UPDATE_WORKFLOW_TEMP_ACTION,
  REMOVE_WORKFLOW_TEMP_ACTION,
  REMOVE_WORKFLOW_ACTION
} from "./constants";

import {
  ByIdState,
  AllIdsState,
  ActionsActions,
  CreateWorkflowTempActionAction,
  ReplaceWorkflowTempActionSuccessAction
} from "./types";

const linkedChildNodeState = (
  state: ByIdState = {},
  action:
    | CreateWorkflowTempActionAction
    | ReplaceWorkflowTempActionSuccessAction
) => {
  const { actions, id } = action;
  const newState = { ...state, ...actions };
  // Set accept/reject action id of new action's parent
  const { assignParent } = actions[id];
  if (assignParent) {
    const { id: parentId, branching } = assignParent;
    const childProp =
      branching === "reject" ? "rejectActionID" : "acceptActionID";
    if (parentId) {
      return set([parentId, childProp], id)(newState);
    }
  }
  return newState;
};

export const byId = (
  state: ByIdState = {},
  action: ActionsActions
): ByIdState => {
  switch (action.type) {
    case LOAD_WORKFLOW_ACTIONS_SUCCESS:
      return { ...state, ...action.actions };
    case CREATE_WORKFLOW_TEMP_ACTION: {
      return linkedChildNodeState(state, action);
    }
    case UPDATE_WORKFLOW_TEMP_ACTION:
      return { ...state, ...action.actions };
    // case REPLACE_WORKFLOW_TEMP_ACTION_SUCCESS: {
    //   const newState = dissocPath(["-1"], linkedChildNodeState(state, action));

    //   return { ...newState, ...action.actions };
    // }
    case REMOVE_WORKFLOW_TEMP_ACTION: {
      const parentId = pathOr(0, ["-1", "assignParent", "id"], state);
      const branch = pathOr(0, ["-1", "assignParent", "branching"], state);

      const currentChild = pathOr(
        undefined,
        [parentId, `${branch}ActionID`],
        state
      );

      let changedParentData = state;
      if (currentChild === -1) {
        changedParentData = assocPath(
          [parentId, `${branch}ActionID`],
          undefined,
          state
        );
      }

      return dissocPath(["-1"], changedParentData);
    }
    default:
      return state;
  }
};

export const allIds = (
  state: AllIdsState = [],
  action: ActionsActions
): AllIdsState => {
  switch (action.type) {
    case LOAD_WORKFLOW_ACTIONS_SUCCESS:
      return union(action.ids)(state);
    case CREATE_WORKFLOW_TEMP_ACTION:
      return [...state, action.id];
    // case REPLACE_WORKFLOW_TEMP_ACTION_SUCCESS: {
    //   const newState = state.filter(id => id !== -1);
    //   return [...newState, action.id];
    // }
    case REMOVE_WORKFLOW_TEMP_ACTION: {
      const newState = state.filter(id => id !== -1);
      return newState;
    }
    case REMOVE_WORKFLOW_ACTION:
      return state.filter(id => id !== action.id);
    default:
      return state;
  }
};

export default combineReducers({
  byId,
  allIds
} as any); // TODO: Upgrade redux
