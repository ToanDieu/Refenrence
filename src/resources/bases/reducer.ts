import { combineReducers } from "redux";
import { union } from "lodash/fp";

import {
  LOAD_BASES_SUCCESS,
  UPDATE_BASE_SUCCESS,
  CREATE_BASE_SUCCESS
} from "./constants";
import { ByIdState, AllIdsState, BasesActions } from "./types";

export const byId = (
  state: ByIdState = {},
  action: BasesActions
): ByIdState => {
  switch (action.type) {
    case LOAD_BASES_SUCCESS:
      return { ...state, ...action.bases };
    case CREATE_BASE_SUCCESS:
      return { ...state, ...action.bases };
    case UPDATE_BASE_SUCCESS:
      return { ...state, ...action.bases };
    default:
      return state;
  }
};

export const allIds = (
  state: AllIdsState = [],
  action: BasesActions
): AllIdsState => {
  switch (action.type) {
    case LOAD_BASES_SUCCESS:
      return union(action.ids)(state);
    case CREATE_BASE_SUCCESS:
      return [...state, action.id];
    default:
      return state;
  }
};

export default combineReducers({
  byId,
  allIds
} as any); // TODO: Upgrade redux
