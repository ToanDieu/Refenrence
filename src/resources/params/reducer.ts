import { combineReducers } from "redux";
import { union } from "lodash/fp";

import { LOAD_BASE_PARAMS_SUCCESS } from "./constants";
import { ByIdState, AllIdsState, ParamsActions } from "./types";

export const byId = (
  state: ByIdState = {},
  action: ParamsActions
): ByIdState => {
  switch (action.type) {
    case LOAD_BASE_PARAMS_SUCCESS:
      return { ...state, ...action.params };
    default:
      return state;
  }
};

export const allIds = (
  state: AllIdsState = [],
  action: ParamsActions
): AllIdsState => {
  switch (action.type) {
    case LOAD_BASE_PARAMS_SUCCESS:
      return union(action.ids)(state);
    default:
      return state;
  }
};

export default combineReducers({
  byId,
  allIds
} as any); // TODO: Upgrade redux
