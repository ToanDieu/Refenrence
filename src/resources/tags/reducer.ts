import { combineReducers } from "redux";
import { union } from "lodash/fp";

import { LOAD_TAGS_SUCCESS, CREATE_TAG_SUCCESS } from "./constants";
import { ByIdState, AllIdsState, TagsActions } from "./types";

export const byId = (state: ByIdState = {}, action: TagsActions): ByIdState => {
  switch (action.type) {
    case LOAD_TAGS_SUCCESS:
      return { ...state, ...action.tags };
    case CREATE_TAG_SUCCESS:
      return { ...state, ...action.tags };
    default:
      return state;
  }
};

export const allIds = (
  state: AllIdsState = [],
  action: TagsActions
): AllIdsState => {
  switch (action.type) {
    case LOAD_TAGS_SUCCESS:
      return union(action.ids)(state);
    case CREATE_TAG_SUCCESS:
      return [...state, action.id];
    default:
      return state;
  }
};

export default combineReducers({
  byId,
  allIds
} as any); // TODO: Upgrade redux
