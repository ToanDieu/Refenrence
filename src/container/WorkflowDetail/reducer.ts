import { LOAD_WORKFLOWS_SUCCESS } from "@/resources/workflows/constants";
import { WorkflowsActions } from "@/resources/workflows/types";
import { LOAD_WORKFLOW_ACTIONS_SUCCESS } from "@/resources/actions/constants";
import { ActionsActions } from "@/resources/actions/types";
import { WorkFlowDetailState, WorkFlowDetailActions } from "./types";
import {
  LOAD_WORKFLOWS_PENDING,
  LOAD_WORKFLOWS_ERROR,
  LOAD_WORKFLOW_ACTIONS_ERROR,
  SET_EDIT_MODE,
  CREATE_WORKFLOW_ACTION_ERROR,
  UPDATE_WORKFLOW_ACTION_ERROR,
  SET_LAST_SAVE,
  SET_BASE_ID,
  DELETE_WORKFLOW_ACTION_ERROR
} from "./constants";

export const initialState: WorkFlowDetailState = {
  baseID: undefined,
  loading: false,
  error: undefined,
  isEditMode: true,
  lastSave: undefined,
  activating: false
};

export default function(
  state = initialState,
  action: WorkFlowDetailActions | WorkflowsActions | ActionsActions
): WorkFlowDetailState {
  switch (action.type) {
    case SET_BASE_ID:
      return { ...state, baseID: action.baseID };
    case LOAD_WORKFLOWS_PENDING:
      return {
        ...state,
        loading: true
      };
    // Continue loading if there is at least a workflow
    case LOAD_WORKFLOWS_SUCCESS: {
      return { ...state, loading: action.ids.length > 0 };
    }
    case LOAD_WORKFLOW_ACTIONS_SUCCESS:
      return { ...state, loading: false };
    case SET_EDIT_MODE:
      return { ...state, isEditMode: action.mode };
    case LOAD_WORKFLOWS_ERROR:
    case LOAD_WORKFLOW_ACTIONS_ERROR:
    case CREATE_WORKFLOW_ACTION_ERROR:
    case UPDATE_WORKFLOW_ACTION_ERROR:
    case DELETE_WORKFLOW_ACTION_ERROR:
      return { ...state, error: action.error };
    case SET_LAST_SAVE:
      return { ...state, lastSave: action.time };
    default:
      return state;
  }
}
