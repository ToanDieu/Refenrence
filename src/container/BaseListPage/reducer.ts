import { BaseListPageState, BaseListPageActions } from "./types";
import {
  LOAD_ALL_PENDING,
  LOAD_ALL_SUCCESS,
  LOAD_ALL_ERROR,
  CHANGE_TAG_FILTER
} from "./constants";
import { BasesActions } from "@/resources/bases/types";

export const FILTER_ALL = -1;

export const initialState: BaseListPageState = {
  loading: true,
  error: undefined,
  tagFilter: FILTER_ALL
};

export default function(
  state = initialState,
  action: BaseListPageActions | BasesActions
): BaseListPageState {
  switch (action.type) {
    case LOAD_ALL_PENDING:
      return { ...state, loading: true };
    case LOAD_ALL_SUCCESS:
      return { ...state, loading: false };
    case LOAD_ALL_ERROR:
      return { ...state, error: action.error };
    case CHANGE_TAG_FILTER:
      return { ...state, tagFilter: action.tagID };
    default:
      return state;
  }
}
