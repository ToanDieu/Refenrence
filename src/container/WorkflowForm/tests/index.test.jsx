import React from "react";
import { render, fireEvent } from "react-testing-library";
import { Provider } from "react-redux";
import { initialize, addTranslationForLanguage } from "react-localize-redux";
import { Notification } from "element-react";

import languages from "@/i18n/languages.json";
import english from "@/i18n/en.json";
import { store as appStore } from "@/store/configureStore";
import { WorkflowForm } from "../index";
import { initialForm } from "../reducer";

jest.mock("@/components/HocComponent", () => ({
  ProtectedScopedComponent: ({ children }) => <div>{children}</div>
}));

describe("<WorkflowForm />", () => {
  let store;

  beforeAll(() => {
    store = appStore;
    const defaultLanguage = "en";
    store.dispatch(initialize(languages, { defaultLanguage }));
    store.dispatch(addTranslationForLanguage(english, "en"));
  });

  it("should render the form", () => {
    const translate = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <WorkflowForm translate={translate} form={initialForm} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should handles when create button was clicked", () => {
    const text = "Create";
    const translate = key => (key === "createWorkflow" ? text : "");
    const clickSpy = jest.fn();
    const { getAllByText } = render(
      <Provider store={store}>
        <WorkflowForm
          translate={translate}
          form={initialForm}
          setDialogVisible={clickSpy}
        />
      </Provider>
    );
    fireEvent.click(getAllByText(text)[0]);
    expect(clickSpy).toBeCalledTimes(1);
  });

  it("should handles when submit button was clicked", () => {
    const translate = jest.fn();
    const submitSpy = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <WorkflowForm
          translate={translate}
          form={{ ...initialForm, name: "Test" }}
          createWorkflow={submitSpy}
        />
      </Provider>
    );
    fireEvent.click(container.querySelector("button[type=submit]"));
    expect(submitSpy).toBeCalledTimes(1);
  });

  it("should handles when name input was changed", () => {
    const translate = jest.fn();
    const changeSpy = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <WorkflowForm
          translate={translate}
          form={initialForm}
          changeFormData={changeSpy}
        />
      </Provider>
    );
    const inputs = container.querySelectorAll("input[type=text]");
    fireEvent.change(inputs[0], { target: { value: "Test" } });
    expect(changeSpy).toBeCalledWith("name", "Test");
  });

  it("should handles when description input was changed", () => {
    const translate = jest.fn();
    const changeSpy = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <WorkflowForm
          translate={translate}
          form={initialForm}
          changeFormData={changeSpy}
        />
      </Provider>
    );
    const inputs = container.querySelectorAll("input[type=text]");
    fireEvent.change(inputs[1], { target: { value: "Test" } });
    expect(changeSpy).toBeCalledWith("description", "Test");
  });

  it("should handles when dialog was closed", () => {
    const translate = jest.fn();
    const closeSpy = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <WorkflowForm
          translate={translate}
          form={initialForm}
          setDialogVisible={closeSpy}
        />
      </Provider>
    );
    const closeButton = container.querySelectorAll("button[type=button]")[1];
    fireEvent.click(closeButton);
    expect(closeSpy).toBeCalledWith(false);
  });

  it("should show error notification when re-render within error", () => {
    const translate = jest.fn();
    const spy = jest.spyOn(Notification, "error");
    render(
      <Provider store={store}>
        <WorkflowForm
          translate={translate}
          form={initialForm}
          error={{ message: "Error" }}
        />
      </Provider>
    );
    expect(spy).toHaveBeenCalled();
  });
});
