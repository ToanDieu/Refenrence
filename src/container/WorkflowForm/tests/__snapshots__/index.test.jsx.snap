// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WorkflowForm /> should render the form 1`] = `
.emotion-0 {
  text-align: center;
}

.emotion-0 p {
  margin: 12px;
}

<div>
  <div
    class="emotion-0"
  >
    <p />
    <div>
      <button
        class="el-button el-button--primary"
        type="button"
      >
        <span />
      </button>
    </div>
  </div>
  <div>
    <div
      class="el-dialog__wrapper"
      style="z-index: 1013; display: none;"
    >
      <div
        class="el-dialog el-dialog--tiny"
        style="top: 15%;"
      >
        <div
          class="el-dialog__header"
        >
          <span
            class="el-dialog__title"
          />
          <button
            class="el-dialog__headerbtn"
            type="button"
          >
            <i
              class="el-dialog__close el-icon el-icon-close"
            />
          </button>
        </div>
        <div
          class="el-dialog__body"
        >
          <form
            class="el-form el-form--label-top"
          >
            <div
              class="el-form-item is-required"
            >
              <div
                class="el-form-item__content"
              >
                <div
                  class="el-input"
                >
                  <input
                    autocomplete="off"
                    class="el-input__inner"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="el-form-item"
            >
              <div
                class="el-form-item__content"
              >
                <div
                  class="el-input"
                >
                  <input
                    autocomplete="off"
                    class="el-input__inner"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          class="el-dialog__footer dialog-footer"
        >
          <button
            class="el-button el-button--primary"
            type="submit"
          >
            <span />
          </button>
        </div>
      </div>
    </div>
    <div
      class="v-modal"
      style="z-index: 1012; display: none;"
    />
  </div>
</div>
`;
